Avoid flickering when showing a layer on a painted background for the first time by disabling async image decoding

shallawa opened the pull request:
#### 238ec677929193ee5fa48b504a09fdce55633145
<pre>
Avoid flickering when showing a layer on a painted background for the first time by disabling async image decoding
<a href="https://bugs.webkit.org/show_bug.cgi?id=270330">https://bugs.webkit.org/show_bug.cgi?id=270330</a>
<a href="https://rdar.apple.com/117533495">rdar://117533495</a>

Reviewed by Simon Fraser;

If an image is decoded asynchronously for a sizeForDrawing different from the
current one, a flicker may happen.

To avoid this flicker, decode the image synchronously if it has more than one
RenderElement in the page and the last time it was decoded asynchronously.

* LayoutTests/TestExpectations:
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::destroyDecodedData):
(WebCore::BitmapImage::draw):
(WebCore::BitmapImage::lastDecodingOptions const):
(WebCore::BitmapImage::lastDecodingOptionsForTesting const): Deleted.
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/ImageObserver.h:
(WebCore::ImageObserver::numberOfClients const):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::imageLastDecodingOptions):

Canonical link: <a href="https://commits.webkit.org/276513@main">https://commits.webkit.org/276513@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/603427581416aecb8126297b36f7b88c69e361dd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44857 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23956 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47347 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47511 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40862 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47160 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28010 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21353 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45435 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21009 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/38623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17913 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18423 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39764 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2904 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41086 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40056 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49177 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19825 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16373 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21143 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42583 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9990 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20818 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/loader/FrameLoader.h:
```diff
@@ -486,6 +488,8 @@ class FrameLoader final : public CanMakeCheckedPtr {
 
     URL m_provisionalLoadErrorBeingHandledURL;
 
+    OptionSet<LayoutMilestone> m_milestones;
```
This seems to be unused.

shallawa replied:
This was removed.

smfr commented on Source/WebCore/rendering/RenderBoxModelObject.cpp:
```diff
@@ -321,14 +321,24 @@ DecodingMode RenderBoxModelObject::decodingModeForImageDraw(const Image& image,
+            return false;
+
+        if (auto* scrollableArea = enclosingClippingLayer->scrollableArea())
+            return scrollableArea->isUserScrollInProgress();
```
I'm not sure that "being in a user scroll" is the best criterion. What about animated programmatic scrolls? What about CSS transform animations?

shallawa replied:
This was removed.

smfr commented on Source/WebCore/platform/graphics/ImageObserver.h:
```diff
@@ -54,6 +57,7 @@ class ImageObserver : public RefCounted<ImageObserver>, public CanMakeWeakPtr<Im
     virtual void imageFrameAvailable(const Image&, ImageAnimatingState, const IntRect* changeRect = nullptr, DecodingStatus = DecodingStatus::Invalid) = 0;
     virtual void changedInRect(const Image&, const IntRect* changeRect = nullptr) = 0;
     virtual void scheduleRenderingUpdate(const Image&) = 0;
+    virtual SingleThreadWeakHashSet<CachedImageClient> visibleClientsInViewport(const Image&, const Document&) { return { }; }
```
Rather than return a `SingleThreadWeakHashSet<CachedImageClient>` from these various new functions, you could just call it `likelyVisibleInViewportMoreThanOnce` or something, and do the counting in `CachedImage::visibleClientsInViewport()`.

shallawa replied:
A different API named `clientsIntersectInViewport()` was added. This API returns `bool`.

shallawa replied:
Another approach was implemented. See the commit message.

smfr commented on LayoutTests/TestExpectations:
```diff
@@ -2274,6 +2274,9 @@ fast/images/image-overlay-mutiple-frames.html [ Skip ]
 # These tests don't have to be platform-specific, but they are only implemented on Mac now.
 fast/images/eps-as-image.html [ Skip ]
 
+# Skip this test to avoid flickering due to multiple async image decoding for different sizeForDrawing.
```
Is there a bug tracking this skipping? What do we need to do to unskip?

shallawa replied:
No tests will skipped. A new layout test was added.

smfr commented on Source/WebCore/rendering/RenderBoxModelObject.cpp:
```diff
@@ -327,8 +327,10 @@ DecodingMode RenderBoxModelObject::decodingModeForImageDraw(const Image& image,
-            if (element() && !element()->hasEverPaintedImages())
+            // No image has not been painted in this element yet it will not flicker with previous painting.
+            auto observer = bitmapImage->imageObserver();
+            bool mayOverlapOtherClients = observer && observer->numberOfClients() > 1 && bitmapImage->lastDecodingOptions().decodingMode() == DecodingMode::Asynchronous;
```
`mayOverlapOtherClients` isn't about overlap any more.

shallawa committed:
Avoid flickering when showing a layer on a painted background for the first time by disabling async image decoding
https://bugs.webkit.org/show_bug.cgi?id=270330
rdar://117533495

Reviewed by Simon Fraser;

If an image is decoded asynchronously for a sizeForDrawing different from the
current one, a flicker may happen.

To avoid this flicker, decode the image synchronously if it has more than one
RenderElement in the page and the last time it was decoded asynchronously.

* LayoutTests/TestExpectations:
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::destroyDecodedData):
(WebCore::BitmapImage::draw):
(WebCore::BitmapImage::lastDecodingOptions const):
(WebCore::BitmapImage::lastDecodingOptionsForTesting const): Deleted.
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/ImageObserver.h:
(WebCore::ImageObserver::numberOfClients const):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::imageLastDecodingOptions):

Canonical link: https://commits.webkit.org/276513@main
